docs/architecture/adr-009-evidence-module.md

```diff
@@ -0,0 +1,179 @@
+type implementing the `Evidence` interface may include arbitrary fields such as
+an infraction time. We want the `Evidence` type to remain as flexible as possible.
+
+When submitting evidence to the `x/evidence` module, the concrete type must provide
+the validator's consensus address, which should be known by the `x/slashing`
+module (assuming the infraction is valid), the height at which the infraction
+occurred and the validator's power at same height in which the infraction occurred.
+
+```go
+type Evidence interface {
+  Route() string
+  Type() string
+  String() string
+  ValidateBasic() Error
+
+  // The consensus address of the malicious validator at time of infraction
+  GetConsensusAddress() ConsAddress
+
+  // Height at which the infraction occurred
+  GetHeight() int64
+
+  // The total power of the malicious validator at time of infraction
+  GetValidatorPower() int64
+
+  // The total validator set power at time of infraction
+  GetTotalPower() int64
+}
+```
+
+### Routing & Handling
+
+Each `Evidence` type must map to a specific unique route and be registered with
+the `x/evidence` module. It accomplishes this through the `Router` implementation.
+
+```go
+type Router interface {
+  AddRoute(r string, h Handler)
+  HasRoute(r string) bool
+  GetRoute(path string) Handler
+  Seal()
+}
+```
+
+Upon successful routing through the `x/evidence` module, the `Evidence` type
+is passed through a `Handler`. This `Handler` is responsible for executing all
+corresponding business logic necessary for verifying the evidence as valid. In
+addition, the `Handler` may execute any necessary slashing and potential jailing.
+Since slashing fractions will typically result from some form of static functions,
+allow the `Handler` to do this provides the greatest flexibility. An example could
+be `k * evidence.GetValidatorPower()` where `k` is an on-chain parameter controlled
+by governance. The `Evidence` type should provide all the external information
+necessary in order for the `Handler` to make the necessary state transitions.
+If no error is returned, the `Evidence` is considered valid.
+
+```go
+type Handler func(Context, Evidence) Error
+```
+
+### Submission
+
+`Evidence` is submitted through a `MsgSubmitEvidence` message type which is internally
+handled by the `x/evidence` module's `SubmitEvidence`.
+
+```go
+type MsgSubmitEvidence struct {
+  Evidence
+}
+
+func handleMsgSubmitEvidence(ctx Context, keeper Keeper, msg MsgSubmitEvidence) Result {
+  if err := keeper.SubmitEvidence(ctx, msg.Evidence); err != nil {
+    return err.Result()
+  }
+
+  // emit events...
+
+  return Result{
+    // ...
+  }
+}
+```
+
+The `x/evidence` module's keeper is responsible for matching the `Evidence` against
+the module's router and invoking the corresponding `Handler` which may include
+slashing and jailing the validator. Upon success, the submitted evidence is persisted.
+
+```go
+func (k Keeper) SubmitEvidence(ctx Context, evidence Evidence) Error {
```

Comment:
This function must also be exposed to other modules (e.g. the client module in IBC)